File: backend/geofence/geofence_test.go
package geofence

import (
	"math"
	"testing"

	"fleet360-backend/geo"
)

// A delivery fence in Dubai with a 200m radius.
func deliveryFence() CircleFence {
	label := "Jebel Ali DC"
	return CircleFence{
		ID:      "stop-1",
		Kind:    FenceDelivery,
		Center:  geo.LatLng{Latitude: 25.0, Longitude: 55.0},
		RadiusM: 200,
		Label:   &label,
	}
}

func TestPointInCircle(t *testing.T) {
	f := deliveryFence()

	t.Run("centre is inside", func(t *testing.T) {
		if !PointInCircle(f.Center, f) {
			t.Fatal("centre should be inside the fence")
		}
	})

	t.Run("a few km away is outside", func(t *testing.T) {
		far := geo.LatLng{Latitude: 25.1, Longitude: 55.1}
		if PointInCircle(far, f) {
			t.Fatal("point ~14km away should be outside a 200m fence")
		}
	})
}

func TestDistanceToCircleM(t *testing.T) {
	f := deliveryFence()
	if d := DistanceToCircleM(f.Center, f); d != 0 {
		t.Fatalf("distance from centre should be 0, got %v", d)
	}
	// 1 degree of latitude is ~111km → ~111000m.
	north := geo.LatLng{Latitude: 26.0, Longitude: 55.0}
	d := DistanceToCircleM(north, f)
	if d < 110_000 || d > 112_000 {
		t.Fatalf("expected ~111000m, got %v", d)
	}
}

func TestDistanceToSegmentM(t *testing.T) {
	a := geo.LatLng{Latitude: 25.0, Longitude: 55.0}
	b := geo.LatLng{Latitude: 25.0, Longitude: 55.1} // due east of a

	t.Run("point on the segment is ~0", func(t *testing.T) {
		mid := geo.LatLng{Latitude: 25.0, Longitude: 55.05}
		if d := DistanceToSegmentM(mid, a, b); d > 1 {
			t.Fatalf("point on segment should be ~0m, got %v", d)
		}
	})

	t.Run("point past the end clamps to the endpoint", func(t *testing.T) {
		// Well east of b — nearest point is b itself.
		past := geo.LatLng{Latitude: 25.0, Longitude: 55.2}
		dSeg := DistanceToSegmentM(past, a, b)
		dToB := geo.HaversineKm(past, b) * 1000
		if math.Abs(dSeg-dToB) > 1 {
			t.Fatalf("past-end distance %v should equal distance to b %v", dSeg, dToB)
		}
	})

	t.Run("degenerate segment falls back to point distance", func(t *testing.T) {
		p := geo.LatLng{Latitude: 25.01, Longitude: 55.0}
		dSeg := DistanceToSegmentM(p, a, a)
		dToA := geo.HaversineKm(p, a) * 1000
		if math.Abs(dSeg-dToA) > 1 {
			t.Fatalf("degenerate segment distance %v should equal distance to a %v", dSeg, dToA)
		}
	})
}

func TestDistanceToPolylineM(t *testing.T) {
	t.Run("empty polyline is +Inf", func(t *testing.T) {
		if d := DistanceToPolylineM(geo.LatLng{}, nil); !math.IsInf(d, 1) {
			t.Fatalf("empty polyline should be +Inf, got %v", d)
		}
	})

	t.Run("single point is haversine distance", func(t *testing.T) {
		pt := geo.LatLng{Latitude: 25.0, Longitude: 55.0}
		line := []geo.LatLng{{Latitude: 25.01, Longitude: 55.0}}
		want := geo.HaversineKm(pt, line[0]) * 1000
		if d := DistanceToPolylineM(pt, line); math.Abs(d-want) > 1 {
			t.Fatalf("single-point polyline distance %v should equal %v", d, want)
		}
	})

	t.Run("takes the nearest segment", func(t *testing.T) {
		// An L-shaped path; the test point sits next to the first leg.
		line := []geo.LatLng{
			{Latitude: 25.0, Longitude: 55.0},
			{Latitude: 25.0, Longitude: 55.1},
			{Latitude: 25.1, Longitude: 55.1},
		}
		pt := geo.LatLng{Latitude: 25.001, Longitude: 55.05} // ~111m north of first leg
		d := DistanceToPolylineM(pt, line)
		if d < 80 || d > 140 {
			t.Fatalf("expected ~111m to nearest leg, got %v", d)
		}
	})
}

func TestWithinCorridor(t *testing.T) {
	line := []geo.LatLng{
		{Latitude: 25.0, Longitude: 55.0},
		{Latitude: 25.0, Longitude: 55.1},
	}

	t.Run("empty polyline is always within", func(t *testing.T) {
		if !WithinCorridor(geo.LatLng{Latitude: 1, Longitude: 1}, CorridorFence{WidthM: 100}) {
			t.Fatal("no corridor defined should mean never off-route")
		}
	})

	t.Run("on the line is within a wide corridor", func(t *testing.T) {
		c := CorridorFence{Polyline: line, WidthM: 500}
		if !WithinCorridor(geo.LatLng{Latitude: 25.0, Longitude: 55.05}, c) {
			t.Fatal("on-line point should be within corridor")
		}
	})

	t.Run("far perpendicular is outside a narrow corridor", func(t *testing.T) {
		c := CorridorFence{Polyline: line, WidthM: 100}
		// ~1.1km north of the line.
		if WithinCorridor(geo.LatLng{Latitude: 25.01, Longitude: 55.05}, c) {
			t.Fatal("1km-off point should be outside a 100m corridor")
		}
	})
}

func TestEvaluateGeofencesCircle(t *testing.T) {
	f := deliveryFence()
	inside := f.Center
	outside := geo.LatLng{Latitude: 25.1, Longitude: 55.1}

	t.Run("ENTER when prev absent and curr inside", func(t *testing.T) {
		ev := EvaluateGeofences(EvaluateInput{Curr: inside, Circles: []CircleFence{f}})
		if len(ev) != 1 || ev[0].Type != EventEnter {
			t.Fatalf("expected single ENTER, got %+v", ev)
		}
		if ev[0].FenceID != "stop-1" || ev[0].FenceKind != FenceDelivery {
			t.Fatalf("ENTER should carry fence identity, got %+v", ev[0])
		}
	})

	t.Run("ENTER when crossing in from outside", func(t *testing.T) {
		ev := EvaluateGeofences(EvaluateInput{Curr: inside, Prev: &outside, Circles: []CircleFence{f}})
		if len(ev) != 1 || ev[0].Type != EventEnter {
			t.Fatalf("expected ENTER, got %+v", ev)
		}
	})

	t.Run("EXIT when crossing out", func(t *testing.T) {
		ev := EvaluateGeofences(EvaluateInput{Curr: outside, Prev: &inside, Circles: []CircleFence{f}})
		if len(ev) != 1 || ev[0].Type != EventExit {
			t.Fatalf("expected EXIT, got %+v", ev)
		}
	})

	t.Run("no event when staying inside", func(t *testing.T) {
		ev := EvaluateGeofences(EvaluateInput{Curr: inside, Prev: &inside, Circles: []CircleFence{f}})
		if len(ev) != 0 {
			t.Fatalf("expected no event for unchanged state, got %+v", ev)
		}
	})

	t.Run("no event when staying outside", func(t *testing.T) {
		ev := EvaluateGeofences(EvaluateInput{Curr: outside, Prev: &outside, Circles: []CircleFence{f}})
		if len(ev) != 0 {
			t.Fatalf("expected no event, got %+v", ev)
		}
	})
}

func TestEvaluateGeofencesCorridor(t *testing.T) {
	line := []geo.LatLng{
		{Latitude: 25.0, Longitude: 55.0},
		{Latitude: 25.0, Longitude: 55.1},
	}
	corridor := &CorridorFence{Polyline: line, WidthM: 100}
	onRoute := geo.LatLng{Latitude: 25.0, Longitude: 55.05}
	offRoute := geo.LatLng{Latitude: 25.02, Longitude: 55.05} // ~2.2km north

	t.Run("DEVIATION when prev on-route and curr off", func(t *testing.T) {
		ev := EvaluateGeofences(EvaluateInput{Curr: offRoute, Prev: &onRoute, Corridor: corridor})
		if len(ev) != 1 || ev[0].Type != EventDeviation {
			t.Fatalf("expected DEVIATION, got %+v", ev)
		}
		if ev[0].OffCorridorM <= 0 {
			t.Fatalf("DEVIATION should report off-corridor distance, got %+v", ev[0])
		}
	})

	t.Run("DEVIATION on first ping already off-route", func(t *testing.T) {
		ev := EvaluateGeofences(EvaluateInput{Curr: offRoute, Corridor: corridor})
		if len(ev) != 1 || ev[0].Type != EventDeviation {
			t.Fatalf("expected DEVIATION for absent prev, got %+v", ev)
		}
	})

	t.Run("RETURN when coming back on-route", func(t *testing.T) {
		ev := EvaluateGeofences(EvaluateInput{Curr: onRoute, Prev: &offRoute, Corridor: corridor})
		if len(ev) != 1 || ev[0].Type != EventReturn {
			t.Fatalf("expected RETURN, got %+v", ev)
		}
	})

	t.Run("no event while on-route", func(t *testing.T) {
		ev := EvaluateGeofences(EvaluateInput{Curr: onRoute, Prev: &onRoute, Corridor: corridor})
		if len(ev) != 0 {
			t.Fatalf("expected no event, got %+v", ev)
		}
	})

	t.Run("single-point corridor is ignored", func(t *testing.T) {
		one := &CorridorFence{Polyline: line[:1], WidthM: 100}
		ev := EvaluateGeofences(EvaluateInput{Curr: offRoute, Prev: &onRoute, Corridor: one})
		if len(ev) != 0 {
			t.Fatalf("corridor with <2 points should produce no events, got %+v", ev)
		}
	})
}

func TestEventMappers(t *testing.T) {
	enter := Event{Type: EventEnter, FenceKind: FencePickup}
	if got := EventTitle(enter, "SHP-1"); got != "SHP-1 arrived at pickup" {
		t.Fatalf("unexpected ENTER title: %q", got)
	}
	if EventSeverity(enter) != "LOW" {
		t.Fatal("ENTER should be LOW severity")
	}
	if got := EventTypeCode(enter); got != "GEOFENCE_ARRIVED_PICKUP" {
		t.Fatalf("unexpected ENTER code: %q", got)
	}

	dev := Event{Type: EventDeviation, OffCorridorM: 250}
	if EventSeverity(dev) != "HIGH" {
		t.Fatal("DEVIATION should be HIGH severity")
	}
	if got := EventTypeCode(dev); got != "GEOFENCE_ROUTE_DEVIATION" {
		t.Fatalf("unexpected DEVIATION code: %q", got)
	}
}

func TestCircleToPolygon(t *testing.T) {
	center := geo.LatLng{Latitude: 25.0, Longitude: 55.0}
	ring := CircleToPolygon(center, 200, 16)

	if len(ring) != 17 {
		t.Fatalf("expected 16 segments + closing point = 17, got %d", len(ring))
	}
	// Closed ring: first and last point coincide.
	if ring[0] != ring[len(ring)-1] {
		t.Fatalf("ring should be closed, first %v last %v", ring[0], ring[len(ring)-1])
	}
	// Each vertex should be ~200m from the centre.
	for i, pt := range ring {
		v := geo.LatLng{Latitude: pt[1], Longitude: pt[0]} // [lng, lat] order
		d := geo.HaversineKm(center, v) * 1000
		if d < 195 || d > 205 {
			t.Fatalf("vertex %d is %vm from centre, expected ~200m", i, d)
		}
	}
}

File: tests/unit/geofence.test.ts
import { describe, expect, it } from 'vitest';
import {
  pointInCircle,
  distanceToCircleM,
  distanceToSegmentM,
  distanceToPolylineM,
  withinCorridor,
  evaluateGeofences,
  geofenceEventType,
  geofenceEventSeverity,
  geofenceEventTitle,
  circleToPolygon,
  type CircleFence,
  type CorridorFence,
} from '@/lib/logistics/geofence';

// Dubai coordinates. ~0.001 deg latitude ≈ 111m.
const STOP = { latitude: 25.2700, longitude: 55.3100 };

const pickupFence: CircleFence = { id: 'pu', kind: 'PICKUP', center: STOP, radiusM: 200, label: 'Jebel Ali' };

// ── circle geometry ────────────────────────────────────────────────────────

describe('pointInCircle / distanceToCircleM', () => {
  it('point at the centre is inside', () => {
    expect(pointInCircle(STOP, pickupFence)).toBe(true);
    expect(distanceToCircleM(STOP, pickupFence)).toBeLessThan(1);
  });

  it('point ~100m away is inside a 200m fence', () => {
    const near = { latitude: 25.2709, longitude: 55.3100 };  // ~100m north
    expect(pointInCircle(near, pickupFence)).toBe(true);
  });

  it('point ~330m away is outside a 200m fence', () => {
    const far = { latitude: 25.2730, longitude: 55.3100 };  // ~330m north
    expect(pointInCircle(far, pickupFence)).toBe(false);
    expect(distanceToCircleM(far, pickupFence)).toBeGreaterThan(200);
  });
});

// ── segment / polyline distance ────────────────────────────────────────────

describe('distanceToSegmentM', () => {
  const a = { latitude: 25.2700, longitude: 55.3100 };
  const b = { latitude: 25.2800, longitude: 55.3100 };  // ~1.1km due north

  it('returns ~0 for a point on the segment', () => {
    const mid = { latitude: 25.2750, longitude: 55.3100 };
    expect(distanceToSegmentM(mid, a, b)).toBeLessThan(2);
  });

  it('returns the perpendicular distance for a point beside the segment', () => {
    // ~100m east of the midpoint
    const beside = { latitude: 25.2750, longitude: 55.3110 };
    const d = distanceToSegmentM(beside, a, b);
    expect(d).toBeGreaterThan(80);
    expect(d).toBeLessThan(120);
  });

  it('clamps to the endpoint for a point beyond the segment end', () => {
    const beyond = { latitude: 25.2900, longitude: 55.3100 };  // ~1.1km past b
    const d = distanceToSegmentM(beyond, a, b);
    // distance to b, not an infinite-line projection
    expect(d).toBeGreaterThan(1000);
    expect(d).toBeLessThan(1300);
  });

  it('handles a degenerate (zero-length) segment', () => {
    const d = distanceToSegmentM({ latitude: 25.2710, longitude: 55.3100 }, a, a);
    expect(d).toBeGreaterThan(100);  // distance to the single point
  });
});

describe('distanceToPolylineM / withinCorridor', () => {
  const polyline = [
    { latitude: 25.2700, longitude: 55.3100 },
    { latitude: 25.2800, longitude: 55.3100 },
    { latitude: 25.2800, longitude: 55.3200 },  // an L-shaped route
  ];
  const corridor: CorridorFence = { polyline, widthM: 150 };

  it('a point on the route is within the corridor', () => {
    expect(withinCorridor({ latitude: 25.2750, longitude: 55.3100 }, corridor)).toBe(true);
  });

  it('a point near the bend is within', () => {
    expect(withinCorridor({ latitude: 25.2800, longitude: 55.3150 }, corridor)).toBe(true);
  });

  it('a point 300m off the route is outside', () => {
    expect(withinCorridor({ latitude: 25.2750, longitude: 55.3130 }, corridor)).toBe(false);
  });

  it('takes the min distance across all segments', () => {
    const d = distanceToPolylineM({ latitude: 25.2800, longitude: 55.3150 }, polyline);
    expect(d).toBeLessThan(150);
  });

  it('no corridor (empty polyline) → always within', () => {
    expect(withinCorridor({ latitude: 0, longitude: 0 }, { polyline: [], widthM: 100 })).toBe(true);
  });
});

// ── transitions ─────────────────────────────────────────────────────────────

describe('evaluateGeofences — circle transitions', () => {
  const outside = { latitude: 25.2740, longitude: 55.3100 };  // ~440m N, outside 200m
  const inside = { latitude: 25.2705, longitude: 55.3100 };   // ~55m N, inside

  it('ENTER when crossing from outside to inside', () => {
    const ev = evaluateGeofences({ prev: outside, curr: inside, circles: [pickupFence] });
    expect(ev).toHaveLength(1);
    expect(ev[0].type).toBe('ENTER');
    expect((ev[0] as { fenceKind: string }).fenceKind).toBe('PICKUP');
  });

  it('EXIT when crossing from inside to outside', () => {
    const ev = evaluateGeofences({ prev: inside, curr: outside, circles: [pickupFence] });
    expect(ev).toHaveLength(1);
    expect(ev[0].type).toBe('EXIT');
  });

  it('no event when staying inside (no re-alert on idle)', () => {
    const stillInside = { latitude: 25.2702, longitude: 55.3100 };
    const ev = evaluateGeofences({ prev: inside, curr: stillInside, circles: [pickupFence] });
    expect(ev).toHaveLength(0);
  });

  it('no event when staying outside', () => {
    const ev = evaluateGeofences({ prev: outside, curr: { latitude: 25.2745, longitude: 55.31 }, circles: [pickupFence] });
    expect(ev).toHaveLength(0);
  });

  it('first ping already inside → ENTER (prev absent = treated as outside)', () => {
    const ev = evaluateGeofences({ prev: null, curr: inside, circles: [pickupFence] });
    expect(ev).toHaveLength(1);
    expect(ev[0].type).toBe('ENTER');
  });

  it('evaluates multiple fences independently', () => {
    const deliveryFence: CircleFence = { id: 'dl', kind: 'DELIVERY', center: { latitude: 25.30, longitude: 55.31 }, radiusM: 200 };
    const ev = evaluateGeofences({ prev: outside, curr: inside, circles: [pickupFence, deliveryFence] });
    // Only the pickup fence transitions
    expect(ev).toHaveLength(1);
    expect((ev[0] as { fenceId: string }).fenceId).toBe('pu');
  });
});

describe('evaluateGeofences — corridor transitions', () => {
  const polyline = [
    { latitude: 25.2700, longitude: 55.3100 },
    { latitude: 25.2800, longitude: 55.3100 },
  ];
  const corridor: CorridorFence = { polyline, widthM: 150 };
  const onRoute = { latitude: 25.2750, longitude: 55.3100 };
  const offRoute = { latitude: 25.2750, longitude: 55.3140 };  // ~400m off

  it('DEVIATION when leaving the corridor', () => {
    const ev = evaluateGeofences({ prev: onRoute, curr: offRoute, circles: [], corridor });
    expect(ev.some(e => e.type === 'DEVIATION')).toBe(true);
    const dev = ev.find(e => e.type === 'DEVIATION') as { offCorridorM: number };
    expect(dev.offCorridorM).toBeGreaterThan(150);
  });

  it('RETURN when coming back onto the corridor', () => {
    const ev = evaluateGeofences({ prev: offRoute, curr: onRoute, circles: [], corridor });
    expect(ev.some(e => e.type === 'RETURN')).toBe(true);
  });

  it('no corridor event while staying on route', () => {
    const ev = evaluateGeofences({ prev: onRoute, curr: { latitude: 25.2760, longitude: 55.31 }, circles: [], corridor });
    expect(ev).toHaveLength(0);
  });

  it('a 1-point corridor is ignored (needs ≥2 for a line)', () => {
    const ev = evaluateGeofences({ prev: onRoute, curr: offRoute, circles: [], corridor: { polyline: [polyline[0]], widthM: 150 } });
    expect(ev).toHaveLength(0);
  });
});

// ── alert mapping helpers ────────────────────────────────────────────────────

describe('geofence event mapping', () => {
  it('maps event types to stable codes', () => {
    expect(geofenceEventType({ type: 'ENTER', fenceId: 'x', fenceKind: 'PICKUP', label: null, distanceM: 10 })).toBe('GEOFENCE_ARRIVED_PICKUP');
    expect(geofenceEventType({ type: 'EXIT', fenceId: 'x', fenceKind: 'DELIVERY', label: null, distanceM: 10 })).toBe('GEOFENCE_DEPARTED_DELIVERY');
    expect(geofenceEventType({ type: 'DEVIATION', offCorridorM: 300 })).toBe('GEOFENCE_ROUTE_DEVIATION');
  });

  it('deviation is HIGH severity, arrivals are LOW', () => {
    expect(geofenceEventSeverity({ type: 'DEVIATION', offCorridorM: 300 })).toBe('HIGH');
    expect(geofenceEventSeverity({ type: 'ENTER', fenceId: 'x', fenceKind: 'PICKUP', label: null, distanceM: 10 })).toBe('LOW');
  });

  it('titles are human-readable and include the shipment number', () => {
    expect(geofenceEventTitle({ type: 'ENTER', fenceId: 'x', fenceKind: 'PICKUP', label: null, distanceM: 10 }, 'LOG-1')).toMatch(/LOG-1.*arrived at pickup/i);
    expect(geofenceEventTitle({ type: 'DEVIATION', offCorridorM: 300 }, 'LOG-1')).toMatch(/LOG-1.*deviated/i);
  });
});

// ── circleToPolygon ──────────────────────────────────────────────────────────

describe('circleToPolygon', () => {
  it('returns a closed ring ([lng,lat], first == last)', () => {
    const ring = circleToPolygon(STOP, 200, 32);
    expect(ring).toHaveLength(33);  // segments + closing point
    expect(ring[0]).toEqual(ring[ring.length - 1]);
  });

  it('every vertex sits ~radius metres from the centre', () => {
    const radiusM = 200;
    const ring = circleToPolygon(STOP, radiusM, 64);
    for (const [lng, lat] of ring) {
      const d = distanceToCircleM({ latitude: lat, longitude: lng }, { id: 'c', kind: 'STOP', center: STOP, radiusM });
      expect(d).toBeGreaterThan(radiusM - 5);
      expect(d).toBeLessThan(radiusM + 5);
    }
  });

  it('a bigger radius produces a wider ring', () => {
    const small = circleToPolygon(STOP, 100, 16);
    const big = circleToPolygon(STOP, 500, 16);
    const spread = (ring: Array<[number, number]>) => Math.max(...ring.map(p => Math.abs(p[1] - STOP.latitude)));
    expect(spread(big)).toBeGreaterThan(spread(small));
  });
});
